### DISCUSSION TITLE
The architecture support scaling out and scaling in on K8s?

### DISCUSSION THREAD

**Discussion by sgfisme** [frePPLe/frepple#460] [Q&A]
Could FrePPLe support the architecture for scaling out and scaling in on K8s?

**Comment by jdetaeye** [comment]
Yes, it's possible.

We already provide a docker container, and a docker compose file.
We have not yet worked on the kubernetes,

So, you can help out&contribute and add your expertise on kubernetes... 
https://github.com/frePPLe/frepple/blob/47807b4e6a3e1f6af408e4fce13187485074983b/doc/installation-guide/docker-container.rst#deployment-with-kubernetes

**Reply by sgfisme** [reply]
Could FrePPle engine support multiple processes to handle the parallel process?
In the current FrePPle architecture, it could leverage the load balancer to handle HTTP requests on the front end. But not sure the engine has the same capability if the single-engine instance has loaded heavily. If yes, we can scale out the pods to add execution resources to improve plan generation efficiency.

**Reply by jdetaeye** [reply]
A few comments on this:

- Where possible the frepple planning engine already uses multiple threads to compute the plans. Technically the plan generation can't be distributed across multiple servers.  In your k8 cluster you'll need to foresee a special node to run the plan generation. That worker node will need sufficient cpu and ram.   (For really big planning problems, we start up a separate server during the daily peak workload and shut it down after that work is done.) 

- In practice the business processes on frepple typically involve a relatively small group of business users.   We are talking max 10s of concurrent users, not 100s or 1000s. In this context we haven't yet had the need to use multiple servers to deal with the front end load.   But technically, it's perfectly possible.

- Finally, for 98% of deployments a single server to do both the front-end and plan generation is sufficient. No need to make it more complex than it needs to be...  
  Are you part of the 2%?   :-)

**Comment by gibhsu** [comment]
Any recommendation we should tune first? I expect it should be CPU instead of memory. Or any indication based on the execution logs or processing time. We monitor the loading on FrePPLe pods and it seems like not use multi-process after I increased the CPU core from 8 to 9.
